File: sort/QuickSort.c
# include <stdio.h>
# include <stdlib.h>
# include <string.h>

typedef struct{
    int r[10];
    int length;
} list;

void swap(list *l,int i,int j){
    int temp;
    temp = l->r[i];
    l->r[i] = l->r[j];
    l->r[j] = temp;
}
void print_heap(list *l){
    int i=0;
    for(i=0;i<=9;i++){
        printf("l->r[%d] = %d\n",i,l->r[i]);
    }
}

int partition(list *l,int low,int high){
    int pivotValue;
    pivotValue = l->r[low];
    while(low<high){
        while(low<high && l->r[high] > pivotValue){
            --high;
        }
        if(low != high){
            swap(l,low,high);
        }
        while(low<high && l->r[low] < pivotValue){
            ++low;
        }
        if(low != high){
            swap(l,low,high);
        }
    }
    //return low; 这里low 和high 相等，都指向当前pivot 都位置
    return high;
}
static void Qsort(list *l,int low,int high){
    int pivot;//枢纽，中间的意思,以此值来划分左右值
    if(low < high){    /* low和high在patition中更改*/
        pivot = partition(l,low,high);
        Qsort(l,1,pivot-1);
        Qsort(l,pivot+1,high);
    }
}
void quickSort(list *l){
    Qsort(l,1,l->length);
}


int main(){
    list *l;
    l = (list*)malloc(sizeof(list));
    l->length = 9;
    int r[] ={0,50,10,90,30,70,40,80,60,20};
    memcpy(l->r,r,sizeof(l->r));
    quickSort(l);
    print_heap(l);
    free(l);
    return 0;
}

File: sort/QuickSort.php
<?php

function swap(&$arr,$i,$j)
{
    list($arr[$i],$arr[$j]) = array($arr[$j],$arr[$i]);
}

function Partition(&$arr,$low,$high)
{
    $pivotValue = $arr[$low];
    while ($low < $high) {
        //一直找到小于$pivotValue的key
        while ($pivotValue <= $arr[$high] && $low < $high) {
            --$high;
        }
        if ($low < $high) {
            swap($arr,$low,$high);
        }
        
        //一直找到大于pivotValue的key
        while ($pivotValue >= $arr[$low] && $low < $high) {
            ++$low;
        }
        
        if ($low < $high) {
            swap($arr,$low,$high);
        }
    }
    //此时$low 和 $high 值一样，代表分割的枢纽所在的key值
    return $high;
}

function Qsort(&$arr,$low,$high)
{
    if ($low < $high) {
        $pivot = Partition($arr,$low,$high);
        Qsort($arr,0,$pivot - 1);
        Qsort($arr,$pivot + 1,$high);
    }
}

function QuickSort(&$arr)
{
    Qsort($arr,0,count($arr) - 1);
}

$arr = [9,5,8,3,7,4,6,2];
QuickSort($arr);
print_r($arr);
/*Array
(
    [0] => 2
    [1] => 3
    [2] => 4
    [3] => 5
    [4] => 6
    [5] => 7
    [6] => 8
    [7] => 9
)*/